pkg/tbtcpg/moving_funds.go

```diff
@@ -0,0 +1,628 @@
+	}
+
+	calculatedHash := crypto.Keccak256(packedWallets)
+
+	if !bytes.Equal(calculatedHash, targetWalletsCommitmentHash[:]) {
+		return nil, ErrWrongCommitmentHash
+	}
+
+	return targetWallets, nil
+}
+
+// GetWalletMembersInfo returns the wallet member IDs based on the provided
+// wallet operator addresses. Additionally, it returns the position of the
+// moving funds task execution operator on the list.
+func (mft *MovingFundsTask) GetWalletMembersInfo(
+	walletOperators []chain.Address,
+	executingOperator chain.Address,
+) ([]uint32, uint32, error) {
+	walletMemberIDs := make([]uint32, 0)
+	walletMemberIndex := 0
+
+	for index, operatorAddress := range walletOperators {
+		operatorID, err := mft.chain.GetOperatorID(operatorAddress)
+		if err != nil {
+			return nil, 0, fmt.Errorf("failed to get operator ID: [%w]", err)
+		}
+
+		// Increment the index by 1 as operator indexing starts at 1, not 0.
+		// This ensures the operator's position is correctly identified in the
+		// range [1, walletOperators.length].
+		if operatorAddress == executingOperator {
+			walletMemberIndex = index + 1
+		}
+
+		walletMemberIDs = append(walletMemberIDs, operatorID)
+	}
+
+	// The task executing operator must always be on the wallet operators list.
+	if walletMemberIndex == 0 {
+		return nil, 0, ErrNoExecutingOperator
+	}
+
+	return walletMemberIDs, uint32(walletMemberIndex), nil
+}
+
+// SubmitMovingFundsCommitment submits the moving funds commitment and waits
+// until the transaction has entered the Ethereum blockchain.
+func (mft *MovingFundsTask) SubmitMovingFundsCommitment(
+	walletPublicKeyHash [20]byte,
+	walletMainUTXO *bitcoin.UnspentTransactionOutput,
+	walletMembersIDs []uint32,
+	walletMemberIndex uint32,
+	targetWallets [][20]byte,
+) error {
+	err := mft.chain.SubmitMovingFundsCommitment(
+		walletPublicKeyHash,
+		*walletMainUTXO,
+		walletMembersIDs,
+		walletMemberIndex,
+		targetWallets,
+	)
+	if err != nil {
+		return fmt.Errorf(
+			"error while submitting moving funds commitment to chain: [%w]",
+			err,
+		)
+	}
+
+	blockCounter, err := mft.chain.BlockCounter()
+	if err != nil {
+		return fmt.Errorf("error getting block counter [%w]", err)
+	}
+
+	currentBlock, err := blockCounter.CurrentBlock()
+	if err != nil {
+		return fmt.Errorf("error getting current block [%w]", err)
+	}
+
+	// To verify the commitment transaction has entered the Ethereum blockchain
+	// check that the commitment hash is not zero.
+	stateCheck := func() (bool, error) {
+		walletData, err := mft.chain.GetWallet(walletPublicKeyHash)
+		if err != nil {
+			return false, err
+		}
+
+		return walletData.MovingFundsTargetWalletsCommitmentHash != [32]byte{}, nil
+	}
+
+	// Wait `5` blocks since the current block and perform the transaction state
+	// check. If the transaction has not entered the blockchain, consider it an
+	// error.
+	result, err := ethereum.WaitForBlockConfirmations(
+		blockCounter,
+		currentBlock,
+		5,
```

Comment:
> What's the reasoning behind 5?
The idea was to wait ~1 minute and check the state.
Changed the code as suggested in 1d487588f893fd2192b6d9c079f07b1361b918aa.

I decided to use the `WaitForBlockHeight` function to implement this:
```
	currentBlock, err := blockCounter.CurrentBlock()
	if err != nil {
		return fmt.Errorf("error getting current block [%w]", err)
	}

	// Make sure the moving funds commitment transaction has been confirmed.
	// Give the transaction at most `6` blocks to enter the blockchain.
	for blockHeight := currentBlock + 1; blockHeight <= currentBlock+6; blockHeight++ {
		err := blockCounter.WaitForBlockHeight(blockHeight)
		if err != nil {
			return fmt.Errorf("error while waiting for block height [%w]", err)
		}

		walletData, err := mft.chain.GetWallet(walletPublicKeyHash)
		if err != nil {
			return fmt.Errorf("error wile getting wallet chain data [%w]", err)
		}

		// To verify the commitment transaction has entered the Ethereum
		// blockchain check that the commitment hash is not zero.
		if walletData.MovingFundsTargetWalletsCommitmentHash != [32]byte{} {
			taskLogger.Infof(
				"the moving funds commitment transaction successfully "+
					"confirmed at block: [%d]",
				blockHeight,
			)
			return nil
		}

		taskLogger.Infof(
			"the moving funds commitment transaction still not confirmed at "+
				"block: [%d]",
			blockHeight,
		)
	}

	taskLogger.Info(
		"failed to verify the moving funds commitment transaction submission",
	)

	return ErrTransactionNotIncluded
```

Alternatively, we could use `WatchBlocks`. It perhaps would be slightly better, since

Reply:
Looks good!